operator/watchers/ciliumendpoint/cilium_endpoint.go

```diff
@@ -0,0 +1,153 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package ciliumendpoint
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"sync"
+
+	"k8s.io/client-go/tools/cache"
+
+	ciliumApiV2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/informer"
+	"github.com/cilium/cilium/pkg/k8s/utils"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+const identityIndex = "identity"
+
+type cepHanlderFunc func(cep *ciliumApiV2.CiliumEndpoint)
+
+var (
+	log      = logging.DefaultLogger.WithField(logfields.LogSubsys, "watchers")
+	errNoCE  = errors.New("object is not a *cilium_api_v2.CiliumEndpoint")
+	indexers = cache.Indexers{
+		cache.NamespaceIndex: cache.MetaNamespaceIndexFunc,
+		identityIndex:        identityIndexFunc,
+	}
+
+	// CiliumEndpointStore contains all CiliumEndpoint present in k8s.
+	// Warning: The CiliumEndpoints stored in the cache are not intended to be
+	// used for Update operations in k8s as some of its fields were are not
+	// populated.
+	CiliumEndpointStore cache.Indexer
+
+	// CiliumEndpointsSynced is closed once the CiliumEndpointStore is synced
+	// with k8s.
+	CiliumEndpointsSynced = make(chan struct{})
+	// once is used to make sure CiliumEndpointsInit is only setup once.
+	once sync.Once
+
+	onUpdateHandlers []cepHanlderFunc
+	onDeleteHandlers []cepHanlderFunc
+)
+
+// identityIndexFunc index identities by ID.
+func identityIndexFunc(obj interface{}) ([]string, error) {
+	switch t := obj.(type) {
+	case *ciliumApiV2.CiliumEndpoint:
+		if t.Status.Identity != nil {
+			id := strconv.FormatInt(t.Status.Identity.ID, 10)
+			return []string{id}, nil
+		}
+		return []string{"0"}, nil
+	}
+	return nil, fmt.Errorf("%w - found %T", errNoCE, obj)
+}
+
+func CiliumEndpointsReinitialize(ctx context.Context, wg *sync.WaitGroup, clientset k8sClient.Clientset) {
+	once = sync.Once{}
+	CiliumEndpointsSynced = make(chan struct{})
+	CiliumEndpointsInit(ctx, wg, clientset)
+}
+
+// CiliumEndpointsInit starts a CiliumEndpointWatcher
+func CiliumEndpointsInit(ctx context.Context, wg *sync.WaitGroup, clientset k8sClient.Clientset) {
+	once.Do(func() {
+		CiliumEndpointStore = cache.NewIndexer(cache.DeletionHandlingMetaNamespaceKeyFunc, indexers)
+		ciliumEndpointInformer := informer.NewInformerWithStore(
```

Comment:
When I started working on the Full CES Reconciliation I was unaware of the hive.
I started by fixing the watchers, implemented full reconciliation and then I learned about the hive and I moved CES Controller to hive.

If I knew about the hive I wouldn't refactor legacy watchers. I can revert this refactoring if needed.
I'll try to revert this refactoring.
This won't help with the main issue - there are still use cases on legacy watchers.

I talked to @pippolo84 about this problem and it's tricky one. It may be hard to move everything out of legacy watchers at once. He agreed that we may have some period of time in which we would have two informers.

I'm trying to move other use case to hive in the #27458
I could work on all remaining use cases but there is already some work done by @pippolo84 so I'm not sure whether it's ok to hijack it.